File: java/src/main/java/com/pqc/mldsa/DsaHash.java
package com.pqc.mldsa;

import com.pqc.common.Keccak;

import static com.pqc.mldsa.DsaField.Q;

/**
 * Hash and sampling functions for ML-DSA (FIPS 204).
 * Uses SHAKE-128/256 via the shared Keccak implementation.
 */
public final class DsaHash {

    private DsaHash() {}

    /** SHAKE-256 hash function H (variable output). */
    public static byte[] h(byte[] input, int outputLen) {
        return Keccak.shake256(input, outputLen);
    }

    /**
     * ExpandA: Generate the k x l matrix A in NTT domain from seed rho.
     * A[i][j] = SampleNTT(SHAKE-128(rho || j || i)).
     * Each element is a polynomial of 256 coefficients in Z_q.
     */
    public static int[][] expandA(byte[] rho, int k, int l) {
        int[][] aHat = new int[k * l][];
        for (int i = 0; i < k; i++) {
            for (int j = 0; j < l; j++) {
                byte[] input = new byte[rho.length + 2];
                System.arraycopy(rho, 0, input, 0, rho.length);
                // FIPS 204: IntegerToBytes(s, 2) where s = 256*i + j => two bytes little-endian
                int s = 256 * i + j;
                input[rho.length] = (byte)(s & 0xFF);
                input[rho.length + 1] = (byte)((s >> 8) & 0xFF);
                aHat[i * l + j] = rejNttPoly(input);
            }
        }
        return aHat;
    }

    /**
     * Rejection sampling of NTT polynomial from XOF output (SHAKE-128).
     */
    private static int[] rejNttPoly(byte[] seed) {
        // Generate enough bytes - 3 bytes yield up to 2 candidates
        // Need 256 coefficients; generate plenty of bytes
        byte[] buf = Keccak.shake128(seed, 256 * 3);
        int[] a = new int[256];
        int j = 0;
        int pos = 0;
        while (j < 256) {
            if (pos + 3 > buf.length) {
                // Need more bytes (very unlikely)
                byte[] more = Keccak.shake128(seed, buf.length * 2);
                buf = more;
                // pos stays the same
            }
            // Extract 3 bytes -> 2 candidates (each < 2^23)
            int b0 = buf[pos] & 0xFF;
            int b1 = buf[pos + 1] & 0xFF;
            int b2 = buf[pos + 2] & 0xFF;
            int d1 = b0 | ((b1 & 0x0F) << 8) | (b2 << 12);
            // Correction: for Dilithium/ML-DSA, coefficients are mod q = 8380417 < 2^23
            // We use CoeffFromThreeBytes: extract 23-bit values
            int val = b0 | (b1 << 8) | ((b2 & 0x7F) << 16);
            if (val < Q) {
                a[j++] = val;
            }
            pos += 3;
        }
        return a;
    }

    /**
     * ExpandS: Generate secret vectors s1 (l polynomials) and s2 (k polynomials)
     * from seed rhoPrime, with coefficients in [-eta, eta].
     * Uses SHAKE-256 with domain separation.
     */
    public static int[][] expandS(byte[] rhoPrime, int k, int l, int eta) {
        int[][] s = new int[k + l][];
        for (int r = 0; r < k + l; r++) {
            byte[] input = new byte[rhoPrime.length + 2];
            System.arraycopy(rhoPrime, 0, input, 0, rhoPrime.length);
            input[rhoPrime.length] = (byte)(r & 0xFF);
            input[rhoPrime.length + 1] = (byte)((r >> 8) & 0xFF);
            s[r] = rejBoundedPoly(input, eta);
        }
        return s;
    }

    /**
     * Rejection bounded polynomial: sample coefficients in [-eta, eta].
     */
    private static int[] rejBoundedPoly(byte[] seed, int eta) {
        int[] a = new int[256];
        byte[] buf = Keccak.shake256(seed, 256 * 2); // plenty of bytes
        int j = 0;
        int pos = 0;
        while (j < 256) {
            if (pos >= buf.length) {
                buf = Keccak.shake256(seed, buf.length * 2);
            }
            int z = buf[pos] & 0xFF;
            int z0 = z & 0x0F;
            int z1 = z >> 4;
            pos++;

            if (eta == 2) {
                if (z0 < 15) {
                    // z0 mod 5 gives value in [0,4], map to [-2,2] by subtracting eta
                    a[j++] = DsaField.modQ(2 - (z0 % 5));
                    if (j >= 256) break;
                }
                if (z1 < 15) {
                    a[j++] = DsaField.modQ(2 - (z1 % 5));
                }
            } else { // eta == 4
                if (z0 < 9) {
                    a[j++] = DsaField.modQ(4 - z0);
                    if (j >= 256) break;
                }
                if (z1 < 9) {
                    a[j++] = DsaField.modQ(4 - z1);
                }
            }
        }
        return a;
    }

    /**
     * ExpandMask: Generate masking vector y from seed and nonce.
     * Each polynomial has coefficients in [-(gamma1-1), gamma1-1].
     */
    public static int[][] expandMask(byte[] rhoPrime, int kappa, int l, int gamma1) {
        int[][] y = new int[l][];
        int gamma1Bits = (gamma1 == 131072) ? 17 : 19; // 2^17 or 2^19
        int polyBytes = 32 * gamma1Bits; // bytes per polynomial
        for (int r = 0; r < l; r++) {
            byte[] input = new byte[rhoPrime.length + 2];
            System.arraycopy(rhoPrime, 0, input, 0, rhoPrime.length);
            int idx = kappa + r;
            input[rhoPrime.length] = (byte)(idx & 0xFF);
            input[rhoPrime.length + 1] = (byte)((idx >> 8) & 0xFF);
            byte[] buf = Keccak.shake256(input, polyBytes);
            y[r] = bitUnpackGamma1(buf, gamma1, gamma1Bits);
        }
        return y;
    }

    /**
     * Unpack polynomial with coefficients in [-(gamma1-1), gamma1].
     */
    private static int[] bitUnpackGamma1(byte[] buf, int gamma1, int bits) {
        int[] r = new int[256];
        for (int i = 0; i < 256; i++) {
            long val = 0;
            int bitStart = i * bits;
            for (int b = 0; b < bits; b++) {
                int bitIdx = bitStart + b;
                int byteIdx = bitIdx / 8;
                int bitOff = bitIdx % 8;
                if (byteIdx < buf.length) {
                    val |= (long)((buf[byteIdx] >> bitOff) & 1) << b;
                }
            }
            // Value is gamma1 - coeff, so coeff = gamma1 - val
            r[i] = DsaField.modQ(gamma1 - (int)val);
        }
        return r;
    }

    /**
     * SampleInBall: Generate challenge polynomial c with exactly tau +/-1 coefficients.
     * Uses SHAKE-256(seed).
     */
    public static int[] sampleInBall(byte[] seed, int tau) {
        byte[] buf = Keccak.shake256(seed, 8 + 256); // 8 bytes for signs + enough for indices
        int[] c = new int[256];

        // First 8 bytes encode signs
        long signs = 0;
        for (int i = 0; i < 8; i++) {
            signs |= ((long)(buf[i] & 0xFF)) << (8 * i);
        }

        int pos = 8;
        for (int i = 256 - tau; i < 256; i++) {
            int j;
            // Sample j uniformly from [0, i]
            while (true) {
                if (pos >= buf.length) {
                    buf = Keccak.shake256(seed, buf.length * 2);
                }
                j = buf[pos++] & 0xFF;
                if (j <= i) break;
            }
            c[i] = c[j];
            c[j] = 1 - 2 * (int)(signs & 1);
            signs >>= 1;
        }
        return c;
    }
}

File: swift/Sources/PQCStandards/MlDsa/DsaHash.swift
// ML-DSA Hash Functions

import Foundation
import CryptoKit

public enum DsaHash {
    /// H (SHAKE-256, variable output)
    public static func h(_ data: [UInt8], outputLen: Int) -> [UInt8] {
        return KemHash.shake256(data, outputLen: outputLen)
    }

    /// Sample in ball: produce a polynomial c with exactly tau +/-1 coefficients
    public static func sampleInBall(seed: [UInt8], tau: Int) -> [Int32] {
        var shake = Shake256()
        shake.absorb(seed)
        var bytes = shake.squeeze(count: 8)

        var signs: UInt64 = 0
        for i in 0..<8 {
            signs |= UInt64(bytes[i]) << (8 * i)
        }

        var c = [Int32](repeating: 0, count: 256)
        for i in (256 - tau)..<256 {
            var j: Int
            repeat {
                let b = shake.squeeze(count: 1)
                j = Int(b[0])
            } while j > i

            c[i] = c[j]
            c[j] = (signs & 1) == 0 ? 1 : DsaField.q - 1
            signs >>= 1
        }
        return c
    }

    /// ExpandA: generate matrix A from seed rho
    public static func expandA(rho: [UInt8], k: Int, l: Int) -> [[[Int32]]] {
        var a = [[[Int32]]]()
        for r in 0..<k {
            var row = [[Int32]]()
            for s in 0..<l {
                row.append(rejNttPoly(rho: rho, r: UInt8(r), s: UInt8(s)))
            }
            a.append(row)
        }
        return a
    }

    /// Rejection sample NTT polynomial from SHAKE-128
    public static func rejNttPoly(rho: [UInt8], r: UInt8, s: UInt8) -> [Int32] {
        var xof = Shake128()
        xof.absorb(rho + [s, r])
        var coeffs = [Int32]()
        coeffs.reserveCapacity(256)
        while coeffs.count < 256 {
            let bytes = xof.squeeze(count: 3)
            var val = Int32(bytes[0]) | (Int32(bytes[1]) << 8) | (Int32(bytes[2]) << 16)
            val &= 0x7FFFFF  // 23 bits
            if val < DsaField.q {
                coeffs.append(val)
            }
        }
        return Array(coeffs.prefix(256))
    }

    /// ExpandS: sample secret vectors s1, s2
    public static func expandS(rhoPrime: [UInt8], k: Int, l: Int, eta: Int) -> ([[Int32]], [[Int32]]) {
        var s1 = [[Int32]]()
        for r in 0..<l {
            s1.append(sampleEtaPoly(seed: rhoPrime, nonce: UInt16(r), eta: eta))
        }
        var s2 = [[Int32]]()
        for r in 0..<k {
            s2.append(sampleEtaPoly(seed: rhoPrime, nonce: UInt16(l + r), eta: eta))
        }
        return (s1, s2)
    }

    /// Sample polynomial with coefficients in [-eta, eta]
    public static func sampleEtaPoly(seed: [UInt8], nonce: UInt16, eta: Int) -> [Int32] {
        let input = seed + [UInt8(nonce & 0xFF), UInt8(nonce >> 8)]
        let bytes = KemHash.shake256(input, outputLen: eta == 2 ? 128 : 136)
        var coeffs = [Int32](repeating: 0, count: 256)

        if eta == 2 {
            for i in 0..<64 {
                let b = bytes[i]
                for j in 0..<2 {
                    let nibble = (b >> (4 * j)) & 0x0F
                    let t0 = Int32(nibble & 3)
                    let t1 = Int32(nibble >> 2)
                    let idx = 4 * i + 2 * j
                    if idx < 256 {
                        coeffs[idx] = DsaField.reduce(Int64(t0 - t1))
                    }
                    if idx + 1 < 256 {
                        // Re-extract for the next
                    }
                }
            }
            // Simpler: process half-bytes
            var ci = 0
            for i in 0..<128 {
                if ci >= 256 { break }
                let b = bytes[i]
                let lo = b & 0x0F
                let t0 = Int32(lo % 5)
                coeffs[ci] = DsaField.reduce(Int64(t0 - Int32(eta)))
                ci += 1
                if ci >= 256 { break }
                let hi = b >> 4
                let t1 = Int32(hi % 5)
                coeffs[ci] = DsaField.reduce(Int64(t1 - Int32(eta)))
                ci += 1
            }
        } else { // eta == 4
            var ci = 0
            for i in 0..<136 {
                if ci >= 256 { break }
                let b = bytes[i]
                let lo = b & 0x0F
                if lo < 9 {
                    coeffs[ci] = DsaField.reduce(Int64(Int32(lo) - 4))
                    ci += 1
                }
                if ci >= 256 { break }
                let hi = b >> 4
                if hi < 9 {
                    coeffs[ci] = DsaField.reduce(Int64(Int32(hi) - 4))
                    ci += 1
                }
            }
        }
        return coeffs
    }

    /// ExpandMask: sample polynomial with coefficients in [-(gamma1-1), gamma1]
    public static func expandMask(rhoPrime: [UInt8], kappa: UInt16, gamma1: Int32) -> [Int32] {
        let input = rhoPrime + [UInt8(kappa & 0xFF), UInt8(kappa >> 8)]
        let gamma1Bits = gamma1 == (1 << 17) ? 18 : 20
        let byteLen = 256 * gamma1Bits / 8
        let bytes = KemHash.shake256(input, outputLen: byteLen)

        var poly = [Int32](repeating: 0, count: 256)
        if gamma1Bits == 18 {
            // 18 bits per coefficient, 4 coefficients per 9 bytes
            for i in 0..<64 {
                let off = 9 * i
                var vals = [Int32](repeating: 0, count: 4)
                let b0 = Int32(bytes[off]); let b1 = Int32(bytes[off+1])
                let b2 = Int32(bytes[off+2]); let b3 = Int32(bytes[off+3])
                let b4 = Int32(bytes[off+4]); let b5 = Int32(bytes[off+5])
                let b6 = Int32(bytes[off+6]); let b7 = Int32(bytes[off+7])
                let b8 = Int32(bytes[off+8])
                vals[0] = b0 | (b1 << 8) | ((b2 & 0x03) << 16)
                vals[1] = (b2 >> 2) | (b3 << 6) | ((b4 & 0x0F) << 14)
                vals[2] = (b4 >> 4) | (b5 << 4) | ((b6 & 0x3F) << 12)
                vals[3] = (b6 >> 6) | (b7 << 2) | (b8 << 10)
                for j in 0..<4 {
                    vals[j] &= ((1 << 18) - 1)
                    poly[4*i + j] = DsaField.reduce(Int64(gamma1 - vals[j]))
                }
            }
        } else {
            // 20 bits per coefficient, 2 coefficients per 5 bytes
            for i in 0..<128 {
                let off = 5 * (i / 2)
                if i % 2 == 0 {
                    let lo = Int32(bytes[off]) | (Int32(bytes[off+1]) << 8) | ((Int32(bytes[off+2]) & 0x0F) << 16)
                    poly[i] = DsaField.reduce(Int64(gamma1 - (lo & ((1 << 20) - 1))))
                } else {
                    let hi = (Int32(bytes[off+2]) >> 4) | (Int32(bytes[off+3]) << 4) | (Int32(bytes[off+4]) << 12)
                    poly[i] = DsaField.reduce(Int64(gamma1 - (hi & ((1 << 20) - 1))))
                }
            }
        }
        return poly
    }
}
